moon/controller.py

```diff
@@ -118,14 +122,38 @@ def send_speed_cmd(self, speed, angular_speed):
             self.virtual_bumper.update_desired_speed(speed, angular_speed)
         self.bus.publish('desired_speed', [round(speed*1000), round(math.degrees(angular_speed)*100)])
 
-    def send_request(self, cmd):
-        """Send ROS Service Request form a single place"""
-        self.publish('request', cmd)
-        while True:
-            dt, channel, data = self.listen()
-            if channel == 'response':
-                return data
-            print(dt, 'ignoring', channel)
+    def on_response(self, timestamp, data):
+        intoken, response = data
+        self.requests[intoken] = {
+            "response": response
+            }
+
+    def poll_responses(self, token, callback):
+        while self.time is None:
+            self.sleep(0.1)
+        request_sent = self.time
+        while self.time - request_sent < timedelta(seconds=5):
+            for i in self.requests:
+                if self.requests[token]["response"] is not None:
+                    response = self.requests[token]["response"]
+                    print(self.time, "controller:response received: token=%s, response=%s" % (token, response))
+                    self.requests.pop(token)
+                    if callback is not None:
+                        callback(response)
+                    return
+            self.sleep(0.1)
+        print(self.time, "controller: send_request timed out: token: %s" % token)
+
+        
+    def send_request(self, cmd, callback=None):
+        """Send ROS Service Request from a single place"""
+        token = hex(self.rand.getrandbits(128))
+        self.requests[token] = {
+            "response": None
+        }
+        print(self.time, "controller:send_request:token: %s, command: %s" % (token, cmd))
+        self.publish('request', [token, cmd])
+        Thread(target=self.poll_responses, args=(token, callback, )).start()
```

Comment:
If that matters, this is not a continuously running thread, this is a worker thread executing a one-time task that either succeeds or times out (within a few seconds) and ends.  If keeping track of threads and waiting with `join` applies even in this scenario, it could probably be added easily as a list of the jobs being worked on already exists. As these terminate themselves quickly, I would assume shutdown notification to each thread may not be necessary albeit that too can be added relatively easily.

Reply:
I posted an updated code with beefed up shutdown handling